### PR TITLE
added: (non-internal) L2 projection support for functions

### PR DESCRIPTION
**Pull request by akva2**
I need this in the Stokes application (to use projected source functions). And since I have separate projection bases there, the current support does not cut it.

@kmokstad I am sure you have some opinion here. I can do it however you see fit, this does the job...

### CODE REVIEW

**Comment by akva2** [comment]
jenkins build this with downstreams please

**Comment by akva2** [comment]
jenkins build this with downstreams please

**Review by kmokstad — CHANGES_REQUESTED** [review]
Github says that `GlbL2projector.h` now is binary, so It's hard to review it efficiently.

image

That said, the general concept looks OK, but I suggest the class name `L2Integrand` instead of the length `IntegrandOrFunction`.

File: src/ASM/GlbL2projector.C [hunk 1]
```diff
@@ -33,6 +34,103 @@ LinAlg::MatrixType GlbL2::MatrixType   = LinAlg::SPARSE;
 LinSolParams*      GlbL2::SolverParams = nullptr;
 
 
+namespace {
+
+static void expandTensorGrid2 (const RealArray* in, RealArray* out)
```
Comment (kmokstad) [inline src/ASM/GlbL2projector.C:39]: Is it necessary with the anonymous `namespace` scoping here? If declaring the functions static, they will be "visible" only for the code in this file in any case (at least, that's what I thought).

Comment (akva2) [reply src/ASM/GlbL2projector.C:39]: static functions is the 'C' way to hide functions. anonymous namespace is the C++ way to do it. just gotten used to doing the latter as of late, but i can of course do it static functions instead if you prefer.

**Comment by akva2** [comment]
So abstract base class named L2Integrand, how about the implementations?

File: src/ASM/GlbL2projector.C (hunk 1 above)
Comment (kmokstad) [reply src/ASM/GlbL2projector.C:39]: OK, but when you do both, like here, it looks like "smør på flesk"/confusing. I have always used the "static" approach elsewhere in the code, so the only reason I can see is consistency, if the result is equivalent. But you can use namespace if you prefer that.

**Comment by kmokstad** [comment]
yea, the names IntegrandNotFunction and FunctionNotIntegrand are perhaps also a bit cumbersome (allthough the meaning is clear). Both classes are integrands in this context, so those names could be confusing. How about changing the base class name to L2IntegrandBase and then L2Integrand and L2Function for the respective sub-classes, or alternatively L2ProbIntegrand and L2FuncIntegrand?

File: src/ASM/GlbL2projector.C (hunk 1 above)
Comment (akva2) [reply src/ASM/GlbL2projector.C:37]: i prefer consistency, so made them static. the reason for the combination is that these are copy-pasted from elsewhere (but i did not want to expose them to global namespace..). updated one, forgot the other.

**Comment by akva2** [comment]
I went for L2Integrand (base class), L2ProbIntegrand and L2FuncIntegrand.

**Comment by akva2** [comment]
jenkins build this with downstreams please

**Comment by akva2** [comment]
jenkins build this with downstreams please

File: src/ASM/GlbL2projector.C [hunk 2]
```diff
@@ -33,6 +34,102 @@ LinAlg::MatrixType GlbL2::MatrixType   = LinAlg::SPARSE;
 LinSolParams*      GlbL2::SolverParams = nullptr;
 
 
+static void expandTensorGrid2 (const RealArray* in, RealArray* out)
```
Comment (kmokstad) [inline src/ASM/GlbL2projector.C:37]: Fine. But you may want to add some small doxy for the these static functions? Or else I assume some jenkins check will complain about it.

File: src/ASM/GlbL2projector.C [hunk 3]
```diff
@@ -33,6 +34,102 @@ LinAlg::MatrixType GlbL2::MatrixType   = LinAlg::SPARSE;
...
+    for (j = 0; j < in[1].size(); j++)
+      for (i = 0; i < in[0].size(); i++, ip++) {
+        out[0][ip] = in[0][i];
+        out[1][ip] = in[1][j];
+        out[2][ip] = in[2][k];
+      }
+}
+
+
+L2ProbIntegrand::L2ProbIntegrand (const ASMbase& ASM, const IntegrandBase& itg) :
```
Comment (kmokstad) [inline src/ASM/GlbL2projector.C:68]: I prefer to use `patch` or `pch` as variable name her for the AMSbase object. ASM is already defined as a namespace elsewhere and could be potentially confusing in the doxy processing (or only in my head...)

File: src/ASM/GlbL2projector.C [hunk 4]
```diff
@@ -33,6 +34,102 @@ LinAlg::MatrixType GlbL2::MatrixType   = LinAlg::SPARSE;
...
+}
+
+
+int L2ProbIntegrand::dim () const
+{
+  return m_itg.getNoFields(2);
+}
+
+
+L2FuncIntegrand::L2FuncIntegrand (const ASMbase& ASM, const FunctionBase& func) :
```
Comment (kmokstad) [inline src/ASM/GlbL2projector.C:86]: same here

File: src/ASM/GlbL2projector.C [hunk 5]
```diff
@@ -33,6 +34,102 @@ LinAlg::MatrixType GlbL2::MatrixType   = LinAlg::SPARSE;
...
+    std::vector<Real> vals = m_func.getValue(X);
+    for (size_t j = 1; j <= m_func.dim(); ++j)
+      sField(j,i+1) = vals[j-1];
+  }
+
+  return true;
+}
+
+
+int L2FuncIntegrand::dim () const
```
Comment (kmokstad) [inline src/ASM/GlbL2projector.C:127]: `size_t`as return type? (since FunctionBase::dim() returns size_t)

File: src/ASM/GlbL2projector.h [hunk 6]
```diff
@@ -29,6 +30,75 @@ typedef std::vector<int>           IntVec;      //!< Vector of integers
...
 
+/*!
+  \brief Abstract class for evaluating integrand or function.
+*/
+
+class L2Integrand {
+public:
+  //! \brief Constructor.
+  //! \param ASM ASM holding geometry to evaluate on
+  L2Integrand(const ASMbase& ASM) : m_asm(ASM) {}
```
Comment (kmokstad) [inline src/ASM/GlbL2projector.h:41]: Here also, I don't think i like ASM as variable name, suggest `patch` or `pch`and then use `myPatch` (or `m_patch` if you prefer) as name on the `const ASMbase&` member

File: src/ASM/GlbL2projector.h [hunk 7]
```diff
@@ -29,6 +30,75 @@ typedef std::vector<int>           IntVec;      //!< Vector of integers
...
+
+  //! \brief Evaluates the entity in a set of points.
+  //! \param sField Matrix with results
+  //! \param gpar Points to evaluate in
+  virtual bool evaluate(Matrix& sField, const RealArray* gpar) const = 0;
+
+  //! \brief Returns dimension of entity to evaluate.
+  virtual int dim() const = 0;
```
Comment (kmokstad) [inline src/ASM/GlbL2projector.h:49]: size_t as return type

File: src/ASM/GlbL2projector.h [hunk 8]
```diff
@@ -29,6 +30,75 @@ typedef std::vector<int>           IntVec;      //!< Vector of integers
...
+/*!
+  \brief Evaluation class for secondary solutions of an integrand.
+*/
+
+class L2ProbIntegrand : public L2Integrand {
+public:
+  //! \brief Constructor.
+  //! \param ASM ASM holding geometry to evaluate on
+  //! \param itg Integrand to evaluate
+  L2ProbIntegrand(const ASMbase& ASM, const IntegrandBase& itg);
```
Comment (kmokstad) [inline src/ASM/GlbL2projector.h:64]: same is in base class

File: src/ASM/GlbL2projector.h [hunk 9]
```diff
@@ -29,6 +30,75 @@ typedef std::vector<int>           IntVec;      //!< Vector of integers
...
+
+  //! \brief Evaluates the secondary solutions in a set of points.
+  //! \param sField Matrix with results
+  //! \param gpar Points to evaluate in
+  bool evaluate(Matrix& sField, const RealArray* gpar) const override;
+
+  //! \brief Returns number of secondary solutions.
+  int dim() const override;
```
Comment (kmokstad) [inline src/ASM/GlbL2projector.h:72]: size_t

File: src/ASM/GlbL2projector.h [hunk 10]
```diff
@@ -29,6 +30,75 @@ typedef std::vector<int>           IntVec;      //!< Vector of integers
...
+/*!
+  \brief Evaluation class for functions.
+*/
+
+class L2FuncIntegrand : public L2Integrand {
+public:
+  //! \brief Constructor.
+  //! \param ASM ASM holding geometry to evaluate on
+  //! \param func Function to evaluate
+  L2FuncIntegrand(const ASMbase& ASM, const FunctionBase& func);
```
Comment (kmokstad) [inline src/ASM/GlbL2projector.h:88]: same

File: src/ASM/GlbL2projector.h [hunk 11]
```diff
@@ -29,6 +30,75 @@ typedef std::vector<int>           IntVec;      //!< Vector of integers
...
+
+  //! \brief Evaluates the function in a set of points.
+  //! \param sField Matrix with results
+  //! \param gpar Points to evaluate in
+  bool evaluate(Matrix& sField, const RealArray* gpar) const override;
+
+  //! \brief Returns number of function components.
+  int dim() const override;
```
Comment (kmokstad) [inline src/ASM/GlbL2projector.h:96]: size_t

**Review by kmokstad — APPROVED** [review]
Went through all now and spotted a few minor things to consider. I don't have full overview of the implications but trust that is is consistent and don't break anything. So you are free to merge at any convenience after resolving/rejecting the smaller things

File: src/ASM/GlbL2projector.C (hunk 2 above)
Comment (akva2) [reply src/ASM/GlbL2projector.C:46]: yeah, copied from originals.

File: src/ASM/GlbL2projector.C (hunk 3 above)
Comment (akva2) [reply src/ASM/GlbL2projector.C]: patch and m_patch.

File: src/ASM/GlbL2projector.C (hunk 5 above)
Comment (akva2) [reply src/ASM/GlbL2projector.C]: changed to size_t

**Comment by akva2** [comment]
jenkins build this with downstreams please

**Comment by akva2** [comment]
jenkins build this with downstreams please

**Pull request merged**